src/providers/mbed_provider/utils.rs

```diff
@@ -219,3 +222,79 @@ pub fn psa_export_public_key_size(key_attrs: &psa_key_attributes_t) -> Result<us
         _ => Err(ResponseStatus::PsaErrorInvalidArgument),
     }
 }
+
+/// Wrapper around raw `psa_key_handle_t` which allows for easier manipulation of
+/// handles and the attributes associated with them.
+pub struct Key<'a>(psa_key_handle_t, &'a Mutex<()>);
+
+impl Key<'_> {
+    /// Create a new key with an empty handle.
+    pub fn new<'a>(key_handle_mutex: &'a Mutex<()>) -> Key<'a> {
+        Key(Default::default(), key_handle_mutex)
+    }
+
+    /// Open a key and store the allocated handle for it.
+    pub fn open_key<'a>(key_id: psa_key_id_t, key_handle_mutex: &'a Mutex<()>) -> Result<Key<'a>> {
+        let mut key_handle: psa_key_handle_t = Default::default();
+        let open_key_status = unsafe {
+            let _guard = key_handle_mutex
+                .lock()
+                .expect("Grabbing key handle mutex failed");
+            psa_crypto_binding::psa_open_key(key_id, &mut key_handle)
+        };
+
+        if open_key_status != PSA_SUCCESS {
+            error!("Open key status: {}", open_key_status);
+            Err(convert_status(open_key_status))
+        } else {
+            Ok(Key(key_handle, key_handle_mutex))
+        }
+    }
+
+    /// Get the attributes associated with the key stored in this handle.
+    pub fn get_attributes(&self) -> Result<psa_crypto_binding::psa_key_attributes_t> {
+        let mut key_attrs = get_empty_key_attributes();
+        let get_attrs_status =
+            unsafe { psa_crypto_binding::psa_get_key_attributes(self.0, &mut key_attrs) };
+
+        if get_attrs_status != PSA_SUCCESS {
+            error!("Get key attributes status: {}", get_attrs_status);
+            Err(convert_status(get_attrs_status))
+        } else {
+            Ok(key_attrs)
+        }
+    }
+
+    /// Release the key stored under this handle.
+    pub fn release_key(&mut self) {
+        unsafe {
+            let _guard = self.1.lock().expect("Grabbing key handle mutex failed");
```

Comment:
I am wondering if there are any implications/safety concerns of putting a safe statement inside an `unsafe` block? Maybe that could be outside of it and still work?

Reply:
I'll have a look through the Nomicon again, see what they think

Reply:
I was thinking that if one of those safe function that we put in an `unsafe` block suddendly become `unsafe`, then we will still execute them without any warning 😕 
But I do not know if that can happen